src/DynamicData/Cache/Internal/ToObservableChangeSet.cs

```diff
@@ -2,107 +2,440 @@
+
+                expirationState.ChangesBuffer.Clear();
+            }
+
+            OnExpirationQueueChanged();
+        }
+
+        private void OnExpirationQueueChanged()
+        {
+            var expirationState = _expirationState!.Value;
+
+            // If there aren't any items queued to expire, check to see if the stream should be terminated (I.E. we just expired the last item).
+            // Otherwise, make sure we have an operation scheduled to handle the next expiration.
+            if (expirationState.Queue.Count is 0)
+            {
+                if (_hasSourceCompleted)
+                    _observer.OnCompleted();
+            }
+            else
+            {
+                // If there's already a scheduled operation, and it doesn't match the current next-item-to-expire time, wipe it out and re-schedule it.
+                var nextExpireAt = expirationState.Queue[0].ExpireAt;
+                if (_scheduledExpiration is ScheduledExpiration scheduledExpiration)
+                {
+                    if (scheduledExpiration.DueTime != nextExpireAt)
+                    {
+                        scheduledExpiration.Cancellation.Dispose();
+                        _scheduledExpiration = null;
+                    }
+                    else
+                    {
+                        return;
+                    }
+                }
+
+                _scheduledExpiration = new()
+                {
+                    Cancellation = _scheduler.Schedule(
+                        state: this,
+                        dueTime: nextExpireAt,
+                        action: static (_, @this) =>
+                        {
+                            lock (@this.SynchronizationGate)
+                            {
+                                @this._scheduledExpiration = null;
+
+                                @this.HandleScheduledExpiration();
+                            }
+
+                            return Disposable.Empty;
+                        }),
+                    DueTime = nextExpireAt
+                };
+            }
+        }
+
+        private void TearDownStates()
+        {
+            _scheduledExpiration?.Cancellation.Dispose();
+            _scheduledExpiration = null;
+
+            _evictionState?.Queue.Clear();
+
+            _expirationState?.Queue.Clear();
+        }
+
+        private readonly EvictionState? _evictionState;
+        private readonly ExpirationState? _expirationState;
+        private readonly Dictionary<TKey, ItemState> _itemStatesByKey;
+        private readonly Func<TObject, TKey> _keySelector;
+        private readonly IObserver<IChangeSet<TObject, TKey>> _observer;
+        private readonly IScheduler _scheduler;
+        private readonly IDisposable _sourceSubscription;
+
+        private bool _hasSourceCompleted;
+        private ScheduledExpiration? _scheduledExpiration;
+
+        private struct ItemState
```

Comment:
Shouldn't a struct always be immutable, and preferably read only?  It would interesting to change these as such and run the benchmarks again,

Reply:
The pitfall with equality of structs is when using them as a `key` within keyed collections, E.G. Dictionary, usually in the interest of creating a composite key. Like you say, they'll *work* but really poorly.

As a rudimentary proof, you can do what I just did and toss `IEquatable<T>` on all those structs and override `.Equals()` and `.GetHashCode()` just to call the base methods. Breakpoints show that none of those methods ever get called in any of the tests.